ARROW-49: [Python] Add Column and Table wrapper interface

xhochy opened the pull request:
After https://github.com/apache/arrow/pull/52 is merged, I'd like to split Column and Table into separate .pyx files, array.pyx seems a bit overcrowded.

wesm commented:
Yep — that can all move to a `table.pyx` file if that works for you

wesm commented on python/pyarrow/array.pyx:
```diff
@@ -360,3 +425,35 @@ cdef class Table:
+    def schema(self):
+        raise NotImplementedError()
+
+    def column(self, index_or_name):
```
I would prefer only to pass indexes here -- retrieval by name can go in a different method. Can you also add a `__getitem__(self, i)` that invokes `self.column(i)`?

wesm commented on python/pyarrow/array.pyx:
```diff
@@ -360,3 +425,35 @@ cdef class Table:
 
         return pd.DataFrame(dict(zip(names, data)), columns=names)
+
+    def name(self):
```
Make this `property name:`

wesm commented on python/pyarrow/array.pyx:
```diff
@@ -287,6 +287,66 @@ cdef class RowBatch:
+        self._check_nullptr()
+        return self.column.null_count()
+
+    def name(self):
```
Property

wesm commented on python/pyarrow/array.pyx:
```diff
@@ -287,6 +287,66 @@ cdef class RowBatch:
+            self._check_nullptr()
+            return (self.length(),)
+
+    def null_count(self):
```
property

wesm commented on python/pyarrow/array.pyx:
```diff
@@ -287,6 +287,66 @@ cdef class RowBatch:
+    def type(self):
+        raise NotImplementedError("Type information not yet implemented")
+
+    def data(self):
```
These should probably also be properties

wesm commented on python/pyarrow/array.pyx:
```diff
@@ -287,6 +287,66 @@ cdef class RowBatch:
+        self._check_nullptr()
+        return self.column.length()
+
+    def length(self):
```
Property maybe? Not as important since we have `__len__`

wesm commented on python/pyarrow/array.pyx:
```diff
@@ -287,6 +287,66 @@ cdef class RowBatch:
+        check_status(pyarrow.ArrowToPandas(self.sp_column, &arr))
+        return pd.Series(<object>arr, name=self.name())
+
+    def _check_nullptr(self):
```
This one can be cdef'd, unless we expect to call it from Python

wesm commented on python/pyarrow/array.pyx:
```diff
@@ -360,3 +425,35 @@ cdef class Table:
+        self._check_nullptr()
+        return self.table.num_columns()
+
+    def num_rows(self):
```
Perhaps make these properties also -- can you add `__len__` that calls `num_rows`?

wesm commented on python/pyarrow/array.pyx:
```diff
@@ -360,3 +425,35 @@ cdef class Table:
+
+    def columns(self):
+        for i in range(self.num_columns()):
+            yield self.column(i)
```
If this returns an iterator, should call the method `itercolumns`.

wesm commented:
Thank you, this is a good start. If you like, feel free to go ahead and move this to `pyarrow.table`

xhochy commented:
Incorporated the comments and also added the missing ChunkedArray wrapper class. Now we should have wrapper for all Arrow classes used in the basic DataFrame conversions.

wesm commented:
+1, thank you, this is excellent
